jep/302/README.adoc

```diff
@@ -0,0 +1,260 @@
+
+This document outlines the design of the snapshotting system we plan to use.
+
+== Specification
+
+_Evergreen_ works with two main components: Jenkins itself, and what we call the evergreen-client.
+
+=== Upgrading/downgrading
+
+Once an evergreen client has been instructed to perform an upgrade, it is responsible for the following operations:
+
+1. (If needed) Initialize the git repository.
+2. Stop Jenkins
+3. Take a snapshot (see <<snapshot>> below)
+5. Perform the instructed upgrade to the given Essentials BOM
+footnote:[Bill Of Materials: this format is currently being designed, but will list everything constituting a version of Essentials: WAR and exact versions of all plugins]
+6. Start new Jenkins version
+7. [[healthcheck]] Check it is responding in a healthy manner.
+8. If not:
+.. Take a snapshot footnote:[this way, if new files were created, we don't just delete them in an unrecoverable way when going back to the previous snapshot].
+.. Rollback to the previous data snapshot and Essentials BOM version.
+To do so, we will generally create a new commit (i.e. avoid `git reset --hard HEAD~`), to keep a durable track of where we went through, accessible through `git log`.
+.. Start (previous) Jenkins version
+// what if starting the previous version doesn't work either?
+9. Report the outcome to the Essentials backend.
+
+// Need to create some diagram to help quickly grok/understand the long list above
+
+[[snapshot]]
+==== Take a snapshot
+
+We use `git` for this purpose, so in this context:
+
+////
+I think .gitignore content must be designed to be able to evolve over time.
+To allow more flexibility, I think the content should be associated between an essentials release/bom to a given .gitignore content.
+////
+
+* Update `.gitignore` content with current Essentials release.
+* `git add .`
+* `git add -u `
+* git commit
++
+[NOTE]
+====
+ The commit log should ideally be made understandable for humans.
+ We will use tags to be able to revert/switch between snapshots in a programmatic reliable way.
+ Each tag name should be designed so that it is clear and easy to link it to a given version of Essentials.
+////
+We need to finish up the work on the BOM to be more precise here.
+////
+====
+
+===== What to backup
+
+In other words, what is going to contain `.gitignore`.
+As said above, this will likely be improved as we go.
+
+.\.gitignore
+[source,gitignore,ffd]
+----
+!plugins
+!war
+----
+
+===== Segregate job configuration and build data
+
+////
+https://issues.jenkins-ci.org/browse/JENKINS-50164 though
+
+This subpart especially is welcoming feedback.
+
+THIS IS A VERY VERY BAD IDEA BECAUSE [...] I'm very interested in :-).
+////
+
+We plan to configure Jenkins to use non default values to put the workspace and builds not under the same location as job configuration itself.
+
+_Workspace Root Directory_:: `${JENKINS_HOME}/workspace/${ITEM_FULLNAME}`
+_Build Record Root Directory_::
+`${JENKINS_HOME}/var/${ITEM_FULL_NAME}/builds`
```

Comment:
`var` looks a little weird.   why was this chosen over something more obvious like `builds`
var implies you can stuff more in here - but there is not enough in the path to disambiguate from builds.